File: frontend/digital-library/src/types/group.ts
// frontend/digital-library/src/types/group.ts

import type { Document, Folder } from "@/types/document";

export type PermissionLevel = "view" | "full";

export interface Workspace {
  id: number;
  name: string;
  description?: string | null;
  owner_id: number;
  default_member_permission: PermissionLevel;
  is_dissolving: boolean;
  dissolve_at?: string | null;
  created_at: string;
  updated_at?: string | null;
}

export interface WorkspaceMember {
  user_id: number;
  username: string;
  full_name?: string | null;
  role: string;
  student_code?: string | null;
  permission_level: PermissionLevel;
  joined_at: string;
  is_owner: boolean;
}

export interface WorkspaceInvitation {
  id: number;
  workspace_id: number;
  workspace_name: string;
  invited_by_name: string;
  message?: string | null;
  permission_level: PermissionLevel;
  created_at: string;
  status: "pending" | "accepted" | "rejected";
}

export interface GroupListItem extends Workspace {
  member_count: number;
  my_permission: PermissionLevel;
  is_owner: boolean;
  last_updated: string;
  role?: string;
}

export interface GroupDocument extends Document {
  owner_id?: number;
  source_document_id?: number | null;
}

export type GroupFolder = Folder;

File: backend/app/schemas/group.py
# backend/app/schemas/group.py

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, ConfigDict, Field


PermissionLevel = Literal["view", "full"]


class GroupCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=150)
    description: Optional[str] = None
    default_member_permission: PermissionLevel = "view"


class GroupUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=150)
    description: Optional[str] = None


class WorkspaceOut(BaseModel):
    id: int
    type: str
    name: str
    description: Optional[str] = None
    owner_id: int
    default_member_permission: PermissionLevel = "view"
    is_deleted: bool = False
    is_dissolving: bool = False
    dissolve_at: Optional[datetime] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class GroupListItem(WorkspaceOut):
    member_count: int
    my_permission: PermissionLevel
    is_owner: bool
    role: str
    last_updated: Optional[datetime] = None


class MemberOut(BaseModel):
    user_id: int
    username: str
    full_name: Optional[str] = None
    role: str
    student_code: Optional[str] = None
    permission_level: PermissionLevel
    joined_at: Optional[datetime] = None
    is_owner: bool


class InviteCreate(BaseModel):
    identifier: str = Field(..., min_length=1, max_length=255)
    message: Optional[str] = Field(default=None, max_length=500)


class InvitationOut(BaseModel):
    id: int
    workspace_id: int
    workspace_name: str
    invited_by_name: str
    message: Optional[str] = None
    permission_level: PermissionLevel
    created_at: Optional[datetime] = None
    status: Literal["pending", "accepted", "rejected"]


class PermissionUpdate(BaseModel):
    permission_level: PermissionLevel


class TransferOwnerPayload(BaseModel):
    new_owner_id: int


class ShareDocumentsPayload(BaseModel):
    document_ids: list[int] = Field(default_factory=list)


class ShareFolderPayload(BaseModel):
    folder_id: int


class ShareResult(BaseModel):
    shared_count: int
